ingest/scripts/compare_hashes.py

```diff
@@ -22,13 +23,102 @@ class Config:
     segmented: str
     nucleotide_sequences: list[str]
     debug_hashes: bool = False
+    slack_hook: str = ""
+
+
+InsdcAccession = str
+JointInsdcAccession = str
+LoculusAccession = str
+Status = str
+
+
+@dataclass
+class SequenceUpdateManager:
+    submit: list[JointInsdcAccession]
+    revise: dict[JointInsdcAccession, LoculusAccession]
+    noop: dict[JointInsdcAccession, LoculusAccession]
+    blocked: dict[Status, dict[JointInsdcAccession, LoculusAccession]]
+    revoke: dict[
+        JointInsdcAccession, dict[LoculusAccession, InsdcAccession]
+    ]  # Map of new grouping joint insdc accessions to map of previous state
+    # i.e. loculus accessions (to be revoked) and their corresponding old joint insdc accessions
+    sampled_out: list[JointInsdcAccession]
+    hashes: list[float]
+
+    def __init__(self, config: Config):
+        self.config = config
+        self.submit = []
+        self.revise = {}
+        self.noop = {}
+        self.blocked = defaultdict(dict)
+        self.revoke = {}
+        self.sampled_out = []
+        self.hashes = []
+
+
+def notify(config: Config, text: str):
+    """Send slack notification with blocked revision details"""
+    if config.slack_hook:
+        requests.post(config.slack_hook, data=json.dumps({"text": text}), timeout=10)
+    logger.warn(text)
 
 
 def md5_float(string: str) -> float:
     """Turn a string randomly but stably into a float between 0 and 1"""
     return int(md5(string.encode(), usedforsecurity=False).hexdigest(), 16) / 16**32
 
 
+def sample_out_hashed_records(
+    insdc_accession_base: str, subsample_fraction: float, sampled_out: dict[str, str], fasta_id: str
+) -> tuple[float, list[dict[str, str]]]:
+    hash_float = md5_float(insdc_accession_base)
+    keep = hash_float <= subsample_fraction
+    if not keep:
+        sampled_out.append({fasta_id: insdc_accession_base, "hash": hash_float})
+    return hash_float, sampled_out
+
+
+def process_hashes(
+    insdc_accession_base: str,
+    fasta_id: str,
+    record: dict,
+    submitted: dict,
+    update_manager: SequenceUpdateManager,
```

Comment:
It's a bit weird that this expects an empty manager (except for config) to be passed in.
Maybe add config as arg instead of update manager and take out config from manager?